File: apps/android/core/network/src/main/kotlin/app/gauja/core/network/DeprecationStore.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.network

import app.gauja.core.model.servers.ProfileId
import javax.inject.Inject
import javax.inject.Singleton
import okhttp3.Headers

data class EndpointDeprecation(
    val endpoint: String,
    val deprecation: String?,
    val sunset: String?,
    val successor: String?,
)

@Singleton
class DeprecationStore @Inject constructor() {
    private val entries = mutableMapOf<ProfileId, LinkedHashMap<String, EndpointDeprecation>>()

    @Synchronized
    fun record(id: ProfileId, endpoint: String, headers: Headers) {
        val link =
            headers
                .values("Link")
                .flatMap { it.split(',') }
                .firstOrNull { it.contains(Regex("rel=\"?successor-version\"?")) }
        if (headers["Deprecation"] == null && headers["Sunset"] == null && link == null) return
        val profile = entries.getOrPut(id) { linkedMapOf() }
        profile[endpoint] =
            EndpointDeprecation(endpoint, headers["Deprecation"], headers["Sunset"], link)
        if (profile.size > 512) profile.remove(profile.keys.first())
    }

    @Synchronized
    fun read(id: ProfileId): List<EndpointDeprecation> = entries[id]?.values?.toList().orEmpty()

    @Synchronized
    fun clear(id: ProfileId) {
        entries.remove(id)
    }
}

File: apps/ios/Packages/Network/Sources/Network/DeprecationStore.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import HTTPTypes
import Model

public struct EndpointDeprecation: Sendable, Equatable {
    public let endpoint: String
    public let deprecation: String?
    public let sunset: String?
    public let successor: String?
}

public actor DeprecationStore {
    private var entries: [ProfileID: [String: EndpointDeprecation]] = [:]
    public init() {}
    func record(_ id: ProfileID, endpoint: String, fields: HTTPFields) {
        let deprecation = fields.first { $0.name.canonicalName == "deprecation" }?.value
        let sunset = fields.first { $0.name.canonicalName == "sunset" }?.value
        let successor = fields.filter { $0.name.canonicalName == "link" }
            .flatMap { $0.value.split(separator: ",").map(String.init) }
            .first { $0.contains("rel=\"successor-version\"") || $0.contains("rel=successor-version") }
        guard deprecation != nil || sunset != nil || successor != nil else { return }
        if entries[id, default: [:]].count >= 512 { entries[id] = [:] }
        entries[id, default: [:]][endpoint] = EndpointDeprecation(
            endpoint: endpoint, deprecation: deprecation, sunset: sunset, successor: successor)
    }
    public func read(_ id: ProfileID) -> [EndpointDeprecation] { Array(entries[id, default: [:]].values) }
    public func clear(_ id: ProfileID) { entries[id] = nil }
}
